### ISSUE TITLE
Implement the user space permissions system

### ISSUE THREAD

**Issue by zalberico** [urbit/urbit#5799]

**Comment by zalberico** [comment]
@drbeefsupreme and @Fang- paired on this to spec out some of the basic things we want.

@drbeefsupreme is taking ownership of this.

**Comment by zalberico** [comment]
Draft spec for this is easier: https://gist.github.com/Fang-/47c90ddd6d33411b6b3ca2c4f0ab56d3 @Fang- and @drbeefsupreme have been working on this and are focused on now building a working MVP.

**Comment by zalberico** [comment]
Update from Mark:
 - Changed permission scope to desk instead of by individual agent
 - There remains a subset of permissions that are still per agent (require more thinking to do per desk)
 - Implemented mark for permissions specification, required for install and optional
 - Soft-blocked on kiln work/app installation flow, Ted, Mark, and Jon to talk to Philip about approach to this (probably build off of their current branch as a hard dependency)
 - For now work on the bits that don't change from Ted and Philip's work.

**Comment by zalberico** [comment]
Updates from Jon:

Last week:
- Mostly a bunch of changing integrating userspace permissions with the agents in 

- Clay branch:
- Added permissions handling to %load
- Added permissions handling to %rein
- Added a couple scries to Clay to get $foam, which is aeon-indexed dome state - - plus non-aeon indexed state like running agents and userspace permissions
- Add %kiln generators for adding/removing userspace perms and the equivalent of Gall's %jolt and %idle, since they're being removed
- Implemented some of the notification system for changing userspace perms via - - %load but am not entirely happy with how agents (as opposed to ducts) are notified at this time
- Rewrote tests for the above as appropriate

This week:
- Continue onwards with userspace perms.
- Figure out the right flow for notifying agents of perm changes
- Add a helper function for agents to determine what permissions they need for their cards
- Start actually enforcing the permission checks and seeing what breaks
- Talk with Mark/Ted to figure out what else comes next

**Comment by zalberico** [comment]
Last week:
 - Mark discussed on fail and on load with Philip, that's sorted
 - User space scrys are in and working

Up Next:
 - Mark to write permissions files for applications in the repo and his own apps
 - Will take notes about any difficult bits with doing this
 - Primary thing this week to also look at kiln integration for this (currently when you install something, gall assumes permissions were granted - kiln needs to block installation if permissions have not been granted).

**Comment by zalberico** [comment]
Last week:
 - More granular permissions, worked with Ted a bit on this
 - Updated the permission files for landscape and current groups - tested this a bit.

This week:
- Writing up a guide for what developers need to do.
- We're getting close to the point where we can communicate this more broadly

One big unknown is how to block install/upgrade of apps that don't have the proper permissions. Has some ideas here, but may follow up with @philipcmonk for advice.

These permissions don't apply to the http interface at all. If you're logged in as the user you can still do anything Eyre lets you do. There's probably a way to fix this (@belisarius222 @philipcmonk), but there are big user/developer experiences there. For this v0 front-end permissions are out of scope, but we want to have some idea for how to handle this.

**Comment by Fang-** [comment]
Draft migration instructions for urbit-dev can be found [here](https://gist.github.com/Fang-/a10ae35c89d69db85f5c531ea40dbd83).

**Comment by Fang-** [comment]
Brief status update:

After discussion with @philipcmonk we realized that the current design for optional permissions isn't sufficient for some cases, unless you want to be brazen and ask for very broad permissions, which would defeat the point of the system. Concretely: if your app cannot know which agents it wants to interact with ahead of time, the current answer is to say "let me interact with all agents", but since that's so broad (notably includes base agents too), you really want to say "let me ask for permission when I know what I want" instead.

So we're pivoting lightly: dropping ahead-of-time optional permissions, and doing "dynamic" permissions instead, adding a mechanism by which agents can request additional permissions during runtime.

The existing mechanism for ahead-of-time _required_ permissions will remain in place unchanged.

As an unrelated simplification step, we'll probably rip out the commit-in-waiting-due-to-permissions mechanism and replace it with a simpler solution that just retries fetching the latest update after permissions change.

@dnbrwstr and I have been hashing out categorization & rendering logic for the permissions. Closing in on something we can actually use. So after the above changes I can take a first stab at frontend integration. Probably do need some treaty changes to be able to display permissions appropriately (getting them pre-install in the first place, and also learning the desk.bill so we know which agents it'll come with).

**Comment by Fang-** [comment]
Status update:
- The "dynamic permissions" described above are in.
- Treaty has been updated to distribute permissions alongside other pre-install metadata.
  - But no backwards compatibility work has been done yet. We'll want some of that, but approximating a clean break here is probably fine. (That is, not being able to hear new/updated non-installed app metadata until you've applied the permissions OTA for both base and landscape.)
- @dnbrwstr and I have continued our adventure into the world of permission rendering. Currently considering the possibility that simply "labeling" permissions with properties is more useful to frontend code than a fully bucket-ized tree structure.
- Irrespective of that I should be able to scaffold the backend-to-frontend-and-back pipeline for permissions. Once that's in a barely-functional place I can hand it off to a frontend person (also @dnbrwstr?) for fleshing out.

**Comment by Fang-** [comment]
Status update:
- @dnbrwstr and I have been going around showing off the UX approach and collecting feedback. We're still narrowing in on the right modality for serving & displaying permissions, I think we're continuing to make progress.
- I started work on making treaty backwards compatible. The compatibility itself isn't that complicated, but the subscription management plumbing is certainly non-trivial. We basically have to do the version negotiation song-and-dance that groups also does. We can't get away with just culling old-style subscriptions, because nacked subscriptions get interpreted as "this software isn't being published anymore", which would interfere with people's Landscape experiences.

I need to spend time this week preparing a presentation for Volcano Summit, which is next week, so progress here might be a little slow.

**Comment by Fang-** [comment]
Current status:
- Treaty backwards compatibility is in. I'm reasonably confident in it, but still needs review & testing, the latter being soft-blocked on:
- Proper upgrade handling in clay still needs to happen. We probably should auto-grant permissions for all apps installed prior to this change. As it stands, permissions are not auto-granted, which blocks the upgrade unless you suspend every installed app, because you cannot grant permissions ahead of time using the old/livenet system.
- Have spoken to @tinnus-napbus and started onboarding him into the codebase/changes here. Will do a prolonged hand-off, get him involved in the backend support for and/or development of the frontend.
